### PR TITLE
Finish M extension

### PR DESCRIPTION
**Pull request by moodlezoup**
Finishes wiring up the M extension to Jolt:
- tracer support h/t @ncitron 
- Adds new constraints
- Fixes various bugs (described in-line)

### CODE REVIEW

File: common/src/rv_trace.rs [hunk 1]
```diff
@@ -355,6 +358,14 @@ impl ELFInstruction {
             RV32IM::VIRTUAL_ASSERT_VALID_DIV0,
         );
 
+        // All instructions in virtual sequence are mapped from the same
+        // ELF address. Thus if an instruction is virtual (and not the last one
+        // in its sequence), then we should *not* update the PC.
+        flags[11] = match self.virtual_sequence_remaining {
+            Some(i) => i != 0,
+            None => false
+        };
```
Comment (moodlezoup) [inline common/src/rv_trace.rs:367]: We need this flag (and so we had to reverse the virtual sequence indexing) for the new non-uniform constraint, which ensures that virtual sequences are executed in full and in order

File: jolt-core/src/field/ark.rs [hunk 2]
```diff
@@ -26,6 +26,23 @@ impl JoltField for ark_bn254::Fr {
         }
     }
 
+    fn to_u64(&self) -> Option<u64> {
```
Comment (moodlezoup) [inline jolt-core/src/field/ark.rs:29]: for debugging

File: jolt-core/src/jolt/instruction/div.rs [hunk 3]
```diff
@@ -190,7 +201,26 @@ impl<const WORD_SIZE: usize> VirtualInstructionSequence for DIVInstruction<WORD_
             advice_value: None,
         });
 
-        virtual_sequence
+        virtual_trace.push(RVTraceRow {
```
Comment (moodlezoup) [inline jolt-core/src/jolt/instruction/div.rs:204]: Previously, we had elided the MOVE instruction as described in the Jolt paper as an optimization, instead writing the advice directly to the destination register
But as it turns out, we actually did need the separate virtual register + MOVE for the edge case where `rd` is equal to `rs1` or `rs2`

File: jolt-core/src/jolt/instruction/lb.rs [hunk 4]
```diff
@@ -55,6 +55,12 @@ impl JoltInstruction for LBInstruction {
                 Box::new(SignExtendSubtable::<F, 8>::new()),
                 SubtableIndices::from(C - 1),
             ),
+            (
```
Comment (moodlezoup) [inline jolt-core/src/jolt/instruction/lb.rs:58]: Fixes bug flagged by the Arithmic team

File: jolt-core/src/jolt/instruction/virtual_assert_lte.rs
Comment (moodlezoup) [inline jolt-core/src/jolt/instruction/virtual_assert_lte.rs:1]: This was implemented as a signed ≤, but it should be an unsigned ≤

File: jolt-core/src/jolt/instruction/sh.rs [hunk 5]
```diff
@@ -31,13 +30,20 @@ impl JoltInstruction for SHInstruction {
...
-            Box::new(IdentitySubtable::<F>::new()),
-            SubtableIndices::from(C - 1),
-        )]
+        vec![
+            (
+                Box::new(IdentitySubtable::<F>::new()),
+                SubtableIndices::from(C - 1),
+            ),
+            (
+                // Not used for lookup, but this implicitly range-checks
```
Comment (sragss) [inline jolt-core/src/jolt/instruction/sh.rs:41]: Let's make a utility function for this so it's self documenting and use across all the memory ops that need range-checking.

`fn RangeCheck(C) -> (Box<Subtable<...>>, SubtableIndices)`

**Review by sragss — APPROVED** [review]
LGTM

File: jolt-core/src/jolt/vm/bytecode.rs [hunk 6]
```diff
@@ -287,8 +309,82 @@ impl<F: JoltField, C: CommitmentScheme<Field = F>> BytecodePolynomials<F, C> {
             DensePolynomial::new(rs2),
             DensePolynomial::new(imm),
         ];
-        let t_read = DensePolynomial::from_usize(&read_cts);
-        let t_final = DensePolynomial::from_usize(&final_cts);
+        let t_read: DensePolynomial<F> = DensePolynomial::from_usize(&read_cts);
+        let t_final: DensePolynomial<F> = DensePolynomial::from_usize(&final_cts);
+
+        #[cfg(test)]
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply jolt-core/src/jolt/vm/bytecode.rs:315]: we need to declare the variables in the same scope as everything else

**Pull request merged**